crates/core/src/dbs/response.rs

```diff
@@ -46,42 +48,99 @@ impl QueryType {
 }
 
 /// The return value when running a query set on the database.
+#[revisioned(revision = 1)]
 #[derive(Debug)]
-pub struct Response {
+pub struct QueryResult {
 	pub time: Duration,
-	pub result: Result<Value>,
+	pub result: Result<Value, DbResultError>,
 	// Record the query type in case processing the response is necessary (such as tracking live
 	// queries).
 	pub query_type: QueryType,
 }
 
-impl Response {
+impl QueryResult {
 	/// Retrieve the response as a normal result
 	pub fn output(self) -> Result<Value> {
-		self.result
+		self.result.map_err(|err| anyhow::anyhow!(err.to_string()))
+	}
+}
+
+impl SurrealValue for QueryResult {
+	fn kind_of() -> Kind {
+		Kind::Object
+	}
+
+	fn is_value(value: &Value) -> bool {
+		matches!(value, Value::Object(_))
 	}
 
 	/// Convert's the response into a value as it is send across the net.
-	pub fn into_value(self) -> Value {
+	fn into_value(self) -> Value {
 		let mut res = Object::new();
-		res.insert("time".to_owned(), format!("{:?}", self.time).into());
+		res.insert("time".to_owned(), Value::String(format!("{:?}", self.time)));
 
 		if !matches!(self.query_type, QueryType::Other) {
-			res.insert("type".to_owned(), self.query_type.to_string().into());
+			res.insert("type".to_owned(), Value::String(self.query_type.to_string()));
 		}
 
 		match self.result {
 			Ok(v) => {
-				res.insert("status".to_owned(), "OK".to_owned().into());
+				res.insert("status".to_owned(), Value::String("OK".to_string()));
 				res.insert("result".to_owned(), v);
 			}
 			Err(e) => {
-				res.insert("status".to_owned(), "ERR".to_owned().into());
-				res.insert("result".to_owned(), e.to_string().into());
+				res.insert("status".to_owned(), Value::String("ERR".to_string()));
+				res.insert("result".to_owned(), Value::String(e.to_string()));
 			}
 		}
 
-		res.into()
+		Value::Object(res)
+	}
+
+	fn from_value(value: Value) -> anyhow::Result<Self> {
```

Comment:
We have a couple methods of this sort. Do you think it's worth adding an internal SurrealValue trait too, with a derive macro xd

Reply:
or perhaps we just use the public one and convert to internal afterwards. Would make life a lot easier and less error prone